mpi4/partitioned_pt2pt: add basic version of partitioned pt2pt

minsii opened the pull request:
## Pull Request Description

This PR implements the AM-based basic version of partitioned pt2pt features that are added in MPI-4 standard.
- CH3 does not support the new feature, simply abort if related functions are called.
- CH4 implements the basic version at CH4 device layer. NM or SHM may overwrite each function to optimize.
- Add basic, dtpool-based, multithreaded partitioned tests in test suites.

### Design overview of the basic version:
- **Message matching**:
    - Because partitioned pt2pt neither matches with regular pt2pt nor collectives, we define separate receive queues for it (see  `MPIDIG_part_rreq_t *part_posted_list; MPIDIG_part_rreq_t *part_unexp_list;` in `MPIDI_CH4_Global_t`).
    - Because partitioned pt2pt matches based on the order of initialization (not by `start` nor `pready|parrived` calls), we handle matching at `MPI_Psend_init|Precv_init` call (by `MPIDIG_PART_SEND_INIT ` AM packet, see [AM protocol] below)

- **Data transfer**: 
    - Receive-buf is ready to accept data after `MPI_Start|Startall`, thus we need notify the send-side once receiver called `start` (by `MPIDIG_PART_CTS` AM packet, see [AM protocol] below)
    - Send-data is ready after `MPI_Pready{_rang,_list}` has been called for each partition. However, this version does not aim for performance. Thus, we transfer data only after all partitions are ready (by `MPIDIG_PART_SEND_DATA` AM packet, see [AM protocol] below). Data may be pipelined by AM framework if size is large.
    - Receive-buf can check completion of each partition by using `MPI_Parrived`. This version does not maintain per-partition completion, it returns TRUE only when all data are arrived (== `rreq` is completed)
    
### AM protocol:
- `MPIDIG_PART_SEND_INIT`: issued by sender at `psend_init`; used for message matching.
- `MPIDIG_PART_CTS`: issued by receiver after (1) the receiver matched the requests and (2) called `start|startall`. Indicates ready-to-receive-data.
- `MPIDIG_PART_SEND_DATA`: issued by sender after (1) all partitions are ready on sender and (2) `MPIDIG_PART_CTS` has been received. Contains entire or partial data (in case pipelined).

### List of major changes/new files

1. Add partitioned pt2pt symbols in CH3 for compilation (all routines simply call MPIR_Assert(0))
b364c943 - ch3: define empty partitioned comm functions and hooks

2. MPIR request changes and new error check routines to support partitioned pt2pt
fbf35943 - mpi/errhan: define partitions API param err check
22b6e82d - mpi/errhan: update req check in start{all}
aba7d39d - mpi/request: add parrived MPIR func
5ca43d4c - mpi/request: extend req test|wait|free for partitioned pt2pt
16fc7815 - mpi/request: define req kinds, fields, utils for partitioned
```
src/include/mpir_request.h ## new request kinds used for partitioned pt2pt
src/mpi/request/mpir_request.c ## changes to MPIR request wait|test|free, new parrived
src/mpi/request/request_impl.c
src/include/mpir_err.h ## new error check routines for partitioned pt2pt
src/mpi/errhan/errnames.txt
```

3. CH4 AM-based generic fallback implementation for partitioned pt2pt (see above design overview)
57fbf22e - ch4: implement req free hook for partitioned pt2pt
bcbd72a8 - ch4/am: implement am generic fallback for partitioned pt2pt
528d7e14 - ch4/am: define error check for incorrect partitioned sync
06518fe8 - ch4/am: define rreq queues for am-based partitioned pt2pt
5cea1622 - ch4/am: define am-specific struct for partitioned req
18f1863d - ch4: define generic partitioned request at device layer
```
src/mpid/ch4/generic/am/mpidig_am_part_utils.h ## utility funcs
/mpid/ch4/generic/am/mpidig_am_part.c ## non-perf-critical partitioned routines (psend|precv_init)
src/mpid/ch4/generic/am/mpidig_am_part.h ## per-critical partitioned routines (start,pready,parrived)
src/mpid/ch4/generic/am/mpidig_am_part_callbacks.c ## AM callback func implementation
src/mpid/ch4/generic/am/mpidig_am_part_callbacks.h ## AM callback func declaration
src/mpid/ch4/generic/am/mpidig_am_init.c ## Register AM callback funcs of partitioned ptp2pt
src/mpid/ch4/generic/am/mpidig_am.h ## Declare new AM packet type for partitioned ptp2pt
src/mpid/ch4/errnames.txt
src/mpid/ch4/include/mpidch4r.h ## Include new header files
src/mpid/ch4/include/mpidpre.h ## Extend request structure for partitioned ptp2pt
src/mpid/ch4/src/ch4_types.h ## Define structures of partitioned ptp2pt AM packets
```

4. Define partitioned routines for CH4 device layer, nm, shm. All fallback to the AM-based generic implementation
018a35c6 - ch4: define MPID interface for partitioned pt2pt
eb0089c9 - ch4: add partitioned pt2pt in am fallback and shm glue layer
744d4883 - ch4/posix: add partitioned pt2pt
78852981 - ch4/ucx: add partitioned pt2pt
46728ad1 - ch4/ofi: add partitioned pt2pt
e6c86f01 - ch4: declare partitioned APIs for shm|nm autogen
```
src/mpid/ch4/ch4_api.txt ## declare autogenerated APIs for shm|nm
src/mpid/ch4/netmod/include/netmod_am_fallback_part.h ## AM-fallback version for perf-critical paths
src/mpid/ch4/netmod/ofi/ofi_part.c ## OFI extension for non-perf-critical partitioned routines (psend|precv_init)
src/mpid/ch4/netmod/ofi/ofi_part.h ## OFI extension for per-critical partitioned routines (start,pready,parrived)
(Similar files are changed or added for POSIX, SHM, UCX)
```

5. Scripts change to add partitioned pt2pt (commit messages are self-explained)
8b2927ae - binding/f08: add choice buffer for psend_init|precv_init
c6f94ea8 - python: extend binding_c.py for partitioned req check
f84ef199 - binding/c: add partitioned pt2pt apis

6. New tests for partitioned pt2pt
5f82536a - test: skip all partitioned tests for ch3 jobs on jenkins
3d0f7960 - test: test match with multiple partitioned calls
0ed51021 - test: add multithreaded partitioned test
cdbde83a - test: add partitioned version of pingping with dtpool
ad7af263 - test: add two sets of basic partitioned tests
```
test/mpi/maint/jenkins/xfail.conf ## skip all partitioned pt2pt tests with ch3 jenkins builds
test/mpi/threads/part/parrived_wait.c ## Test using multiple threads on both sender and receiver
test/mpi/part/multipart.c ## Test message matching with multiple partitioned calls
test/mpi/part/parrived.c ## Test more on the receiver side using parrived
test/mpi/part/start_pready.c ## Test with various pready scenarios
(Skiped Makefile, testlist, configure changes)
```

<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->

## Expected Impact

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

minsii commented:
test:mpich/ch3/most
test:mpich/ch4/most

minsii commented:
test:mpich/ch4/most

minsii commented:
test:mpich/ch3/most

minsii commented:
Should fixed all ch4 failures. TODO: fix test xfail for ch3 builds.

test:mpich/ch4/most

minsii commented:
Solved test xfails for ch3:

test:mpich/ch3/most

minsii commented:
Added a test for message matching with multiple partitioned calls
test:mpich/ch3/most
test:mpich/ch4/most

minsii commented:
Testing again after commits cleanup. TODO: test with GPU 

test:mpich/ch3/most
test:mpich/ch4/most

minsii commented:
@hzhou do you want to review the partitioned pt2pt PR?

All tests passed in the previous push. The latest push was for commit cleanup/adding notes, so I don't expect new failure.

The missing items are: (1) test with GPU (2) PVAR for partitioned pt2pt. But I think we can go ahead and merge the existing PR first. It is already large.

hzhou commented on src/mpid/ch4/include/mpidpre.h:
```diff
@@ -227,6 +232,42 @@ typedef struct MPIDI_prequest {
+    MPIR_Context_id_t context_id;
+    MPI_Datatype datatype;
+    int active_flag;            /* flag indicating whether in a start-complete active period */
+} MPIDI_part_request_t;
```
Usually, the request either go thru native path where use the MPIDI field, or go thru the AM path, where we use the am fields. Your design seems to require both the native part and am part, is this necessary? If possible, try not to create new patterns.

minsii replied:
I think we need both am and device-layer (MPIDI) fields. My reason is that, the partitioned req is more similar to an RMA window, rather than another type of `one-time-used-request`. Specifically,
- We need store a set of persistent attributes such as `buffer`, `partitions`, `count`, for many rounds of communication, similar to RMA mode. These attributes will be used for both AM path and native path. Thus, we put them in `MPIDI` as device-layer info.
- The AM part is used for MPIDIG am protocol. When we develop the native path in future, if we don't use the same protocol for start-complete sync, we may put `am` into a union at that time; likely we may want to simply reuse the AM protocol for sync as they are not in perf-critical path, similar to that of RMA (e.g., AM based sync + RDMA based perf-critical data transfer)

minsii replied:
My understanding is that you are OK if I keep the current request structure and keep all partitioned attributes in `MPIDI_part_request_t`. Can you please confirm?

hzhou replied:
Yes, you can keep it for now. I dont' think it is critical. We can easily move the struct to different place if we want to, right?

hzhou commented on src/mpid/ch4/src/ch4_part.c:
```diff
@@ -0,0 +1,128 @@
+    MPIDI_workq_vci_progress_unsafe();
+
+    *request = part_req_create((void *) buf, partitions, count, datatype, dest, tag, comm,
+                               MPIR_REQUEST_KIND__PART_SEND);
```
Why not let the device create the request as pt2pt and RMA paths do?

minsii replied:
I guess you are suggesting that we let the `MPIDIG` or a native path create the request?

The reason now I create it at the device layer is that, I think they neither netmod nor shmmod nor am dependent. 

Maybe we can do something similar as that of `win_create`. See pseudo code below:
```
#ifdef MPIDI_CH4_DIRECT_NETMOD
    mpi_errno = MPIDI_NM_mpi_psend_init(...);
    MPIR_ERR_CHECK(mpi_errno);
#else
    mpi_errno = MPIDIG_mpi_psend_init(...);
    MPIR_ERR_CHECK(mpi_errno);
#endif
```

In `MPIDIG_mpi_psend_init`:
 (1) creates the request
 (2) set device-layer attributes
 (3) initialize AM resource/protocal
 (4) call `NM|SHM_mpi_psend_init_hook` for any nm|shm specific initialization 

What do you think?

minsii replied:
Whiling doing the refactoring, I feel that we still have to create a new pattern for partitioned pt2pt initialization, rather than completely following `win_create`. I.e., it is pt2pt, thus it might bring more benefit than windows if we allow SHM do optimization and skip generic AM. Thus, I'd keep the current `SHM|NM_psend_init` selection in `MPID_Psend_init`, but only move the request creation part to MPIDIG.

Besides, it is unsure how NM|SHM may use the hooks to do additional initialization. Thus, I'd skip the hook definition for now.

hzhou replied:
I agree. 

Also now I realize the request we created in `psend_init` is not the same as request we create from the device -- it is more like a window. So in that sense, I think it make more sense to have the request creation at higher layer, potentially even at the mpir layer. At which layer do we allocate window?

I agree that we should skip the hook definition now. In general, I think we should delay any design until we need it in a implementation.

minsii replied:
Window is created at device layer. See `MPIDIG_mpi_win_create`->`win_init`

hzhou replied:
I see. That feels a bit odd. I guess we designed that window synchronization only to be implemented in generic layer, so maybe that makes sense to have the window creation there. But frankly, I think that should be done at MPIR-layer, and call device hooks to populate fields if necessary.

With partitioned communication, do we plan to always have matching done via active messages? If yes, then I think we can follow the same pattern as windows since you can't skip AM anyway.

minsii replied:
Can't do window creation at MPIR because the RMA code in ch4 was completely different from that of ch3. Will be messy if we abstract some common parts of creation into MPIR but leave the remaining in device layer.

For partitioned pt2pt, we will need AM-based sync for both the potential RDMA version and the pt2pt version. But for each version, we may need exchange different attributes (e.g., `seq_no` for pt2pt version) and thus we will need different AM packet and request attributes.

Based on this thought, the attributes in `MPIDI_part_request_t` are generic for all approaches; the ones in `MPIDIG_part_request_t am` is AM-version specific. When a different version is added in future, the structure in `MPIDI_part_request_t` will look like
```
typedef struct MPIDI_part_request {
    union {
         MPIDIG_part_request_t am;
         xxx;  <- new version
    } u;

    /* partitioned attributes */
    void *buffer;
    int partitions;
    MPI_Aint count;             /* count per partition */
    int rank;
    int tag;
    MPIR_Context_id_t context_id;
    MPI_Datatype datatype;
} MPIDI_part_request_t;
```

hzhou commented on src/mpid/ch4/src/ch4_part.c:
```diff
@@ -0,0 +1,128 @@
+    MPIDI_av_entry_t *av = MPIDIU_comm_rank_to_av(comm, source);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
+    MPIDI_workq_vci_progress_unsafe();
```
Unless you have specific reason, let's not bother adding the workq progress here. We are not going to add workq for partitioned communication. And workq is being deprecated anyway.

hzhou commented on src/mpid/ch4/src/ch4_part.c:
```diff
@@ -0,0 +1,128 @@
+
+    MPIDI_av_entry_t *av = MPIDIU_comm_rank_to_av(comm, source);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
```
Could you take the vci lock inside device layer as the pt2pt path does. I have #4746 also moving the vci lock down in the RMA path. If you do it here, someone have to come here later to move them down again for extra work.

minsii replied:
I actually don't follow the vci design well. Will make the change as you suggested.

minsii replied:
Btw, I am not sure when we need call `MPIDI_workq_vci_progress` together with the vci lock. Based on #4746 , is it needed only at `MPID` layer when `MPIDI_CH4_USE_WORK_QUEUES` is on?

minsii replied:
One more question regarding vci lock for `psend_init|precv_init`. As I think they are more similar to `win_create` rather than regular pt2pt, I am trying to find where we do lock for `win_create`. But not found in neither `MPID` nor `MPIDIG`. Do we lock such resource allocation calls at MPI layer?

minsii replied:
Right. Then we still need lock for `psend|precv_init` as two threads may call it on the same comm. Thanks for explanation.

hzhou commented on src/mpid/ch4/src/ch4_part.h:
```diff
@@ -0,0 +1,129 @@
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_PART_STARTALL);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
+    MPIDI_workq_vci_progress_unsafe();
```
Same comment regarding vci lock and workq progress, and following functions.

minsii replied:
moved to MPIDIG

hzhou commented on src/mpid/ch4/src/ch4_part.h:
```diff
@@ -0,0 +1,129 @@
+    goto fn_exit;
+}
+
+MPL_STATIC_INLINE_PREFIX int MPID_Pready_range(int low, int high, MPIR_Request * request)
```
The parameter names here are different from the netmod API (`low` vs `partition_low`). I would prefer shorter names and I know the longer names are from MPI spec. But either way, I think we should keep them consistent unless with specific reasons.

minsii replied:
Will change to be consistent with the spec names.

hzhou commented on src/mpid/ch4/src/ch4_part.c:
```diff
@@ -0,0 +1,128 @@
+    *request = part_req_create((void *) buf, partitions, count, datatype, dest, tag, comm,
+                               MPIR_REQUEST_KIND__PART_SEND);
+#ifdef MPIDI_CH4_DIRECT_NETMOD
+    mpi_errno = MPIDI_NM_mpi_psend_init(buf, partitions, count, datatype, dest, tag, comm,
```
By the way, I think you should split this commit to separate ch4-part and mpidig part. The flow is `ch4 -> netmod -> mpidig`, so mpidig are not the same component level as the ch4 level here.

minsii replied:
The reason I put all ch4-layer code in a single commit is that, a part of the requests initialization is in `ch4_part.c`. Separating them will make the initialization logic incomplete. But as in the previous comment, we are discussing to move all initialization to `MPIDIG`. I may separate the ch4-part into a separate commit.

hzhou commented on src/mpid/ch4/include/mpidch4.h:
```diff
@@ -71,6 +71,10 @@ MPL_STATIC_INLINE_PREFIX void MPID_Request_set_completed(MPIR_Request *) MPL_STA
+MPL_STATIC_INLINE_PREFIX int MPID_Part_request_is_active(MPIR_Request *) MPL_STATIC_INLINE_SUFFIX;
+MPL_STATIC_INLINE_PREFIX void MPID_Part_request_free_hook(MPIR_Request *) MPL_STATIC_INLINE_SUFFIX;
+MPL_STATIC_INLINE_PREFIX void MPID_Part_request_complete_hook(MPIR_Request *)
+    MPL_STATIC_INLINE_SUFFIX;
```
Could you add in commit message where is this hook being called?

minsii replied:
done

hzhou commented on src/include/mpir_request.h:
```diff
@@ -310,6 +318,9 @@ static inline int MPIR_Request_is_active(MPIR_Request * req_ptr)
                 return (req_ptr)->u.persist.real_request != NULL;
+            case MPIR_REQUEST_KIND__PART_SEND:
+            case MPIR_REQUEST_KIND__PART_RECV:
+                return MPID_Part_request_is_active(req_ptr);
```
I don't quite understand why this (`MPID_Part_request_is_active`) need be device function call. Can't we simply use a atomic flag?

Could you clarify when is the request considered active? Is it between `MPI_Start` and `MPI_Test`? If it is, then it is entirely a MPIR-layer flag.

minsii replied:
Yes, it is active between `start`-`test|wait`. I think how to implement the "active" status is device-internal, e.g., for persistent req, it is based on whether there is an active req. So I put it inside device.

Having said that, I don't see any stronger reason to hide it inside device. I am OK to move it to `MPIR`, if it simplifies the code.

minsii replied:
moved `active_flag` to MPIR_Request; thus deleted `MPID_Part_request_is_active` and `MPID_Part_request_complete_hook`

hzhou commented on src/mpi/request/mpir_request.c:
```diff
@@ -168,6 +168,15 @@ int MPIR_Request_completion_processing(MPIR_Request * request_ptr, MPI_Status *
+        case MPIR_REQUEST_KIND__PART_SEND:
+        case MPIR_REQUEST_KIND__PART_RECV:
+            {
+                MPID_Part_request_complete_hook(request_ptr);
```
I am not sure about the goal of this hook either.

minsii replied:
It inactivates the req. But if we are going to move the `active` flag to `MPIR`, we can delete the hook. Will do.

hzhou commented on src/mpi/request/mpir_request.c:
```diff
@@ -168,6 +168,15 @@ int MPIR_Request_completion_processing(MPIR_Request * request_ptr, MPI_Status *
+                MPID_Part_request_complete_hook(request_ptr);
+
+                MPIR_Request_extract_status(request_ptr, status);
+                mpi_errno = request_ptr->status.MPI_ERROR;
```
I am also surprised that the completion process of partitioned request is different from persistent request. Can you clarify the difference, specifically, why we can't simply reuse the persistent request path here?

minsii replied:
The persistent pt2pt relies on an internal `real_request` which was created when sending or receiving the message at `MPID_Start` by calling `MPID_Irecv|Isend`. It does not expose the chance for NM or SHM native optimization. Thus, when completing a persistent request, it always frees `real_request` at MPIR.

We don't want to reuse the model of persistent calls for partitioned pt2pt because we allow NM or SHM to overwrite with different data transfer approaches. Currently basic version creates only a single AM req to track data transfer, but when we parallelize each partition transfer in future, we may need multiple reqs to keep track on concurrent data transfers. Thus, I don't think it is a good idea to force the same request completion process at MPIR for partitioned pt2pt.

hzhou reviewed (commented):
Just reviewed the first a few commits. Let's resolve and clarify before continuing.

minsii commented on src/include/mpi.h.in:
```diff
@@ -895,7 +895,11 @@ typedef int (MPIX_Grequest_wait_function)(int, void **, double, MPI_Status *);
+#define MPIX_ERR_PARTITION            MPICH_ERR_FIRST_MPIX+6 /* Invalid partition value */
+#define MPIX_ERR_PARTITIONS           MPICH_ERR_FIRST_MPIX+7 /* Invalid partitions value */
+#define MPIX_ERR_MISMATCH_SIZE        MPICH_ERR_FIRST_MPIX+8 /* Mismatch buffer size in partitioned pt2pt */
+
```
I agree that the error codes might be a bit fine-grained. I am not very confident with the error code definition as no error code is standardized yet for partitioned comm. Thus, I was just listing all possible errors **explicitly mentioned** in the standard. For other **implementation-dependent** error, I treat them as internal error.

I am trying to find an existing code that might have a similar scenario.  However, what I see is that the existing codes actually follow a specific/fine-grained way. E.g., see below
```
#define MPI_ERR_INFO        28      /* */
#define MPI_ERR_INFO_KEY    29      /* */
#define MPI_ERR_INFO_VALUE  30      /* */
#define MPI_ERR_INFO_NOKEY  31      /* */
```

I list them as `MPIX_` because of two reasons (1) they are not yet standardized (2) no clear view how the error codes will be standardized. I am not sure how much impact it will be if we change the `MPIX_` error code to `MPI_` (or delete `MPIX_` error code) later once they are standardized in spec, but sounds like less impact than changing an `MPI_` error code.

hzhou replied:
> I agree that the error codes might be a bit fine-grained. I am not very confident with the error code definition as no error code is standardized yet for partitioned comm. Thus, I was just listing all possible errors **explicitly mentioned** in the standard. For other **implementation-dependent** error, I treat them as internal error.
> 
> I am trying to find an existing code that might have a similar scenario. However, what I see is that the existing codes actually follow a specific/fine-grained way. E.g., see below
> 
> ```
> #define MPI_ERR_INFO        28      /* */
> #define MPI_ERR_INFO_KEY    29      /* */
> #define MPI_ERR_INFO_VALUE  30      /* */
> #define MPI_ERR_INFO_NOKEY  31      /* */
> ```
> 
> I list them as `MPIX_` because of two reasons (1) they are not yet standardized (2) no clear view how the error codes will be standardized. I am not sure how much impact it will be if we change the `MPIX_` error code to `MPI_` (or delete `MPIX_` error code) later once they are standardized in spec, but sounds like less impact than changing an `MPI_` error code.

I see that all MPI error class are explicitly listed in the standard, that makes me think that we should not add additiional error class unless it is listed in the standard.

The argument for not to use `MPIX_ERR_` class is of independent reason. `MPIX_` are extension, that in theory, we can remove without affecting the main functionality. Partitioned communication is the officical standard feature, so you can't remove those MPIX_ error codes as if they are extensions.

Can we do this PR without add extra error classes, for example,  use `MPI_ERR_OTHER` if necessary?

minsii replied:
Yea, we can use `MPI_ERR_OTHER` for now. Easier for us to change when the partitioned errors are well standardized.

minsii replied:
Will make the changes

minsii commented on src/mpi/errhan/errnames.txt:
```diff
@@ -229,6 +229,15 @@ MPI_TYPECLASS_INTEGER, or MPI_TYPECLASS_COMPLEX
 A memory object id was expected but a non-memory object id was passed instead
 
+**mismatchsize: Partitioned send and receive buﬀers are not identical in size
+**mismatchsize %d %d %d %d: Partitioned send and receive buﬀers are not identical in size; \
```
Will change to `partmismatchsize`

minsii commented on src/mpi/errhan/errnames.txt:
```diff
@@ -229,6 +229,15 @@ MPI_TYPECLASS_INTEGER, or MPI_TYPECLASS_COMPLEX
+**mismatchsize %d %d %d %d: Partitioned send and receive buﬀers are not identical in size; \
+%d bytes received but receive buffer size is %d
+**partitions: Negative partitions
+**partitions %d: Negative partitions, value is %d
```
Will change to `**partitionsneg`, to follow `**countneg:Negative count`

minsii commented on src/mpi/errhan/errnames.txt:
```diff
@@ -229,6 +229,15 @@ MPI_TYPECLASS_INTEGER, or MPI_TYPECLASS_COMPLEX
+**partitions %d: Negative partitions, value is %d
+**partition: Invalid partition
+**partition %d: Invalid partition, value is %d 
+**partition %d %d: Invalid partition range, values are %d %d
```
Will change to `**partitioninvalid`. I try to do not shorten to `part` when unneeded, as it might be ambiguous about partition v.s. partitions

hzhou commented:
In the RMA case, `put/get` may be issued via workq, so the actual `put/get` may not get posted before we call workq progress, and since RMA sync need have those put/get posted, that is the reason we call workq_progress before a lot of RMA calls.

In this case, partitioned functions are never issues via workq, there is nothing in partitioned local matching require workq progress, so there isn’t a need to call workq progress.


Btw, I am not sure when we need call MPIDI_workq_vci_progress together with the vci lock. Based on #4746<https://github.com/pmodels/mpich/pull/4746> , is it needed only at MPID layer when MPIDI_CH4_USE_WORK_QUEUES is on?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pmodels/mpich/pull/5033#discussion_r568964567>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AALNM7QZ5VQFT3TTXF6YPTDS5BZUBANCNFSM4WV66QZQ>.

minsii commented on src/mpi/errhan/baseerrnames.txt:
```diff
@@ -82,3 +82,6 @@ MPI_ERR_RMA_FLAVOR  58      **rmaflavor
 MPIX_ERR_NOREQ      61      **nomemreq
+MPIX_ERR_PARTITION  62      **partition
+MPIX_ERR_PARTITIONS 63      **partitions
+MPIX_ERR_MISMATCH_SIZE 64   **mismatchsize
```
All replaced with `MPI_ERR_OTHER `

minsii commented:
Testing after some refactoring work. Working on separation of the CH4 commit.

test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
I'll take another review. By the way, you can but you don't always need to clear the full test suite before requesting another round of review.

hzhou reviewed (commented):
I'll continue tomorrow.

minsii commented on src/mpid/ch4/generic/am/mpidig_am_part_utils.h:
```diff
@@ -0,0 +1,190 @@
+        MPIDIG_REQUEST(sreq, req->part_am_req.part_req_ptr) = part_sreq;
+
+        /* count of entire send buffer; partitions are contiguous in memory. */
+        MPI_Count count = MPIDI_PART_REQUEST(part_sreq, count) * part_sreq->u.part.partitions;
```
`count` in `psend|precv_init` is `MPI_Count`. If we cast to `MPI_Aint`, I am afraid that issue might be caused by integer overflow.

minsii replied:
Emm, just noticed `MPIDI_SHM_am_isend` (where the `count` will be passed in) uses `MPI_Count` too.

hzhou replied:
If you worry, add an `MPIR_Assert`. We can't really handle counts that is larger than `MPIR_AINT_MAX`, so even if you take `MPI_Count` here, there will bound to have issues at other parts.

minsii replied:
Is there any plan to change `MPIDI_SHM_am_isend ` to use `MPI_Aint`? Otherwise, it sounds wrong to me to do a twice casting.

hzhou replied:
If `MPIDI_SHM_am_isend` is not using `MPI_Aint`, that need be fixed.

minsii replied:
Alright. Let me change it to aint + assert

minsii replied:
done

minsii commented on src/mpid/ch4/generic/am/mpidig_am_part.h:
```diff
@@ -0,0 +1,131 @@
+    MPIR_Assert(MPIR_cc_get(MPIDIG_PART_REQUEST(sreq, u.sreq).ready_cntr) <=
+                sreq->u.part.partitions);
+
+    mpi_errno = MPIDIG_part_issue_data(sreq, is_local);
```
I check conditions inside `MPIDIG_part_issue_data` as the same logic applies for both `pready` functions. I see your confusion. I can move the "wait for all partitions ready" check back to pready functions.

minsii replied:
done

minsii commented on src/mpid/ch4/generic/am/mpidig_am_part.h:
```diff
@@ -0,0 +1,131 @@
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDIG_MPI_PREADY_RANGE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIG_MPI_PREADY_RANGE);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
```
Sure. I didn't pay attention to granularity of locks in the first version. I will change for both `pready` functions.

minsii replied:
done

hzhou commented:
> I'll continue tomorrow.

Looks like all the rest of the commits are straight forward, so I don't expect much issues. The only commit that is difficult to review is " ch4: implement basic am-based partitioned pt2pt". It seems okay, but there are too many components inside that makes one uncertain. If you could split it into finer components -- e.g. API interface, Datastructure, AM frame workd, AM implementation -- that will help.

minsii commented on src/mpid/ch4/generic/am/mpidig_am_part.h:
```diff
@@ -0,0 +1,131 @@
+    MPIR_Assert(MPIR_cc_get(MPIDIG_PART_REQUEST(sreq, u.sreq).ready_cntr) <=
+                sreq->u.part.partitions);
+
+    mpi_errno = MPIDIG_part_issue_data(sreq, is_local);
```
done

minsii commented:
Let me try to split them a bit after cleaning up the remaining issues. But it really might make sense only for review. As one cannot revert only partial commits of the CH4 AM part.

hzhou commented:
> Let me try to split them a bit after cleaning up the remaining issues. But it really might make sense only for review. As one cannot revert only partial commits of the CH4 AM part.

When we come back later trying to optimize it or implement new features, we may check back to understand the design. Havine each components separate helps that understanding. If we have difficulty digest the commit now, we'll have the same difficulty later.

minsii commented:
All comments should have been addressed. @hzhou can you please check again? 

The original ch4 am huge commit is now divided to multiple pieces with detailed design notes in commit message (top commit comes later)
bb325305 - ch4/am: implement am generic fallback for partitioned pt2pt
07a11cbb - ch4/am: define error check for incorrect partitioned sync
79ea4b14 - ch4/am: define rreq queues for am-based partitioned pt2pt
841b14be - ch4/am: define am-specific struct for partitioned req
57f6a953 - ch4: define generic partitioned request at device layer

After the ch4 am part, the later commits are also reordered so that each commit can be compiled (top commit comes later):
**(MPI-layer)**
83b46607 - binding/f08: add choice buffer for psend_init|precv_init
23f31c13 - python: extend binding_c.py for partitioned req check
5a90d8aa - binding/c: add partitioned comm apis
**(MPIR-layer)**
f07b19e9 - mpi/errhan: define partitions API param err check
b7789554 - mpi/errhan: update req check in start{all}
21b2ae57 - mpi/request: add parrived MPIR func
69e2d140 - mpi/request: extend req test|wait|free for partitioned pt2pt
**(MPID-layer)**
b3a92db5 - ch4: implement req free hook for partitioned pt2pt
1d1c045b - ch4: define MPID interface for partitioned pt2pt
**(SHM|NM-layer which calls CH4 AM)**
f9dd9322 - ch4/shm: add partitioned comm in shm
4b69c45f - ch4/shm: add partitioned comm in shm am
cbab9d6b - ch4/posix: add partitioned comm
54d6ef2a - ch4/ucx: add partitioned comm
93d6035b - ch4/ofi: add partitioned comm
a23bef92 - ch4/nm: add partitioned comm in netmod am
f9528b32 - ch4: declare partitioned APIs for shm|nm autogen

minsii commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented on src/mpid/ch4/include/mpidpre.h:
```diff
@@ -227,6 +227,17 @@ typedef struct MPIDI_prequest {
+    MPI_Aint count;             /* count per partition */
+    int rank;
+    int tag;
+    MPIR_Context_id_t context_id;
```
The `comm` should be stored in the request already, right? (If not, it should). Why do we need to store `context_id` here? It is redundant.

minsii replied:
You are right that we don't need `context_id` for request created by `psend_init|precv_init`, since `comm` is already in `MPIR_Request`. But we need it for unexpected rreq which is temporarily created when the receiver receives the message matching AM (see  `MPIDIG_part_send_init_target_msg_cb`)

hzhou replied:
> But we need it for unexpected rreq which is temporarily created when the receiver receives the message matching AM (see `MPIDIG_part_send_init_target_msg_cb`)

I see. We look up the `root_comm` when we receive the unexpected am message, right? So you can just set the comm to `root_comm`, right?

minsii replied:
Reverted the change, we need `context_id` set for both posted_rreq and unexp_rreq so that both types can be matched with the same routine `MPIDIG_part_dequeue`.

I don't use `root_comm` in partitioned pt2pt am matching. They are separate queues from the regular pt2pt queues anyway. I feel that involving `root_comm` just makes the logic complex unnecessarily. Simply compare context_id is more easy to understand.

hzhou commented on src/mpid/ch4/generic/am/mpidig_am_part_utils.h:
```diff
@@ -0,0 +1,31 @@
+ * Updated as atomic cntr because AM callback and local operation may be
+ * concurrently performed by threads. */
+#define MPIDIG_PART_REQ_UNSET 0 /* Initial status set at psend|precv_init */
+#define MPIDIG_PART_REQ_CTS 2   /* Indicating receiver is ready to receive data
```
Even it isn't used, I think we should define `1`. It just read weird with the gap

minsii replied:
Defined `MPIDIG_PART_REQ_PARTIAL_ACTIVATED` and added note to explain the unused intermediate status.

hzhou commented on src/mpid/ch4/generic/am/mpidig_am_part_utils.h:
```diff
@@ -0,0 +1,31 @@
+ * Called when receiver matches request, sender receives CTS, or either side calls start. */
+MPL_STATIC_INLINE_PREFIX int MPIDIG_PART_REQ_INC_FETCH_STATUS(MPIR_Request * part_req)
+{
+    int prev_stat = 0;
```
The setting to `0` is unnecessary.

minsii replied:
Changed to `int prev_stat = MPL_atomic_fetch_add_int(&MPIDIG_PART_REQUEST(part_req, status), 1);`

hzhou commented on src/mpid/ch4/include/mpidpre.h:
```diff
@@ -227,7 +227,31 @@ typedef struct MPIDI_prequest {
+    MPIR_Request *peer_req_ptr;
+    union {
+        MPIDIG_part_sreq_t sreq;
+        MPIDIG_part_rreq_t rreq;
```
In too many places, `sreq` and `rreq` are of type `MPIR_Request *`, so the names here are a source of confusion. I think you can just use `send` and `recv` here.

minsii replied:
Sure. will replace

hzhou commented on src/mpid/ch4/include/mpidpre.h:
```diff
@@ -227,7 +227,31 @@ typedef struct MPIDI_prequest {
+
 typedef struct MPIDI_part_request {
+    MPIDIG_part_request_t am;
+
```
Could you remind me again why this `am` cannot be a part of the AM union as the other AM requests (e.g. `(req)->dev.ch4.am.part`)?

minsii replied:
This `am` is in `(req)->dev.ch4.part_req` which is union with `(req)->dev.ch4.am`. 

`(req)->dev.ch4.part_req` is used by the "persistent" request created at `psend|precv_init`, whereas `(req)->dev.ch4.am` is used by the "one-time" request created at each AM-based data transfer.

We also define an extended type in `(req)->dev.ch4.am` for partitioned AM data transfer, see `MPIDIG_REQUEST(rreq, req->part_am_req)`. It stores the pointer of `(req)->dev.ch4.part_req` so that we can inactivate `part_req` when the "one-time" AM request completes (see `part_send_data_target_cmpl_cb` and `MPIDIG_part_send_data_origin_cb`)

hzhou commented on src/mpid/ch4/netmod/include/netmod_am_fallback.h:
```diff
@@ -11,5 +11,6 @@
 #include "netmod_am_fallback_probe.h"
 #include "netmod_am_fallback_rma.h"
 #include "netmod_am_fallback_coll.h"
+#include "netmod_am_fallback_part.h"
```
The commit message should be "ch4/nm: add partitioned comm in netmod am *fallback*

minsii replied:
done

hzhou commented on src/mpid/ch4/shm/src/Makefile.mk:
```diff
@@ -12,6 +12,7 @@ noinst_HEADERS += src/mpid/ch4/shm/src/shm_impl.h    \
         src/mpid/ch4/shm/src/shm_am_fallback_recv.h  \
         src/mpid/ch4/shm/src/shm_am_fallback_rma.h   \
         src/mpid/ch4/shm/src/shm_am_fallback_send.h  \
+        src/mpid/ch4/shm/src/shm_am_fallback_part.h  \
```
I think this commit, "ch4/shm: add partitioned comm in shm am " should be squashed with the `netmod` am fallback commit. They are doing the same thing and always need to be updated together anyway.

minsii replied:
done

hzhou reviewed (commented):
Thanks for splitting the design commits, they are much easier to understand now. I left a few comments.

minsii commented:
test:mpich/ch3/most
test:mpich/ch4/most

minsii commented:
@hzhou I have addressed all comments. Can you please check again?

minsii commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented on test/mpi/threads/part/parrived_wait.c:
```diff
@@ -0,0 +1,247 @@
+
+/* Utility functions */
+#define VAL(_partition, _count, _off, _iter)        \
+        (double) ((_partition) * (_count) + (_off) + (_iter))   /* Equals to `index + iter` */
```
Suggest rid of the macro.

The reason that this macro is bad is that it introduces 4 parameters whose meaning can only be inferred from the context.

minsii replied:
replaced with `VAR(index, iter)` similar to previous tests.

hzhou commented on test/mpi/threads/part/testlist:
```diff
@@ -0,0 +1,2 @@
+parrived_wait 2 arg=-spartitions=8 arg=-rpartitions=8 arg=-tot_count=64 arg=-iteration=5
+parrived_wait 2 arg=-spartitions=16 arg=-rpartitions=32 arg=-tot_count=64 arg=-iteration=5
```
Last line needs a newline

minsii replied:
fixed

hzhou reviewed (commented):
Reviewed the tests

minsii commented on test/mpi/part/Makefile.am:
```diff
@@ -0,0 +1,39 @@
+start_pready_list_wait_SOURCES  = start_pready.c
+
+parrived_wait_CPPFLAGS = -DTEST_WAIT $(AM_CPPFLAGS)
+parrived_wait_SOURCES  = parrived.c
```
I see a terminator $ at the last line when enabling :set line with vim. Could you please share with me your method to detect the issue?

hzhou replied:
I was just looking at github per-commit changes. It shows a red circle if the last line is missing a newline. 

I suspect some of the notes are confusing because I was looking at per-commit changes and they disapper when the next commits combines. If if is too difficult for you to locate the offending commits, you can ignore my notes. Just make sure the final files are all terminated.

minsii replied:
Added a new line at end

minsii commented on test/mpi/part/parrived.c:
```diff
@@ -0,0 +1,179 @@
+    int errs = 0;
+    for (int i = 0; i < rcount; i++) {
+        double exp_val = VAL(rp, rcount, i, iter);
+        if (buf[rp * rcount + i] != exp_val) {
```
When writing past tests, I have seen similar bugs several time that the value assignment is not consistent in data set and validation subroutines. Thus, I use a single macro here to ensure it always gives the same value with given inputs.

I am fine to delete the macro to save discussion time. It is trivial to me.

hzhou replied:
I guess I am fine if the macros is `VAL(index, iter)`, since that is clear on the meaning even read out-of-context.

minsii replied:
replaced with `VAL(index, iter)`

minsii commented on test/mpi/part/parrived.c:
```diff
@@ -0,0 +1,179 @@
+                MTestPrintfMsg(1, "Rank %d sends partition %d with count %ld of basic elements\n",
+                               rank, i, scount);
+                MPI_Pready(i, &req);
+            }
```
out-of-order should be fine. E.g., `pready(partiton2); pready(partiton1);`

minsii replied:
added a pready-out-of-order version in `parrived.c`. See test `parrived_pready_out_of_order`

minsii commented on test/mpi/part/parrived.c:
```diff
@@ -0,0 +1,179 @@
+
+            for (int j = 0; j < rpartitions; j += 2) {
+                /* mimic receive-side computation that handles two partitions differently. */
+                int part1_complete = 0, part2_complete = 0;
```
It is copied from standard. Fine to rename the variables.

minsii replied:
replaced

minsii commented on test/mpi/part/parrived.c:
```diff
@@ -0,0 +1,179 @@
+    if (rank == 0) {
+        MPI_Psend_init(buf, spartitions, 1, send_type, 1, 0, MPI_COMM_WORLD, MPI_INFO_NULL, &req);
+
+        for (int x = 0; x < iteration; x++) {
```
good suggestion. replaced with `iter`

minsii commented on test/mpi/part/parrived.c:
```diff
@@ -0,0 +1,179 @@
+            reset_buf();
+            MPI_Start(&req);
+
+            for (int j = 0; j < rpartitions; j += 2) {
```
I copied from the standard example. Makes sense to replace with `i`.will do.

minsii commented on test/mpi/part/parrived.c:
```diff
@@ -0,0 +1,179 @@
+#else
+            do {
+                MPI_Test(&req, &flag, MPI_STATUS_IGNORE);
+            } while (!flag);
```
makes sense. replaced to 
```
            MPI_Test(&req, &flag, MPI_STATUS_IGNORE);
            if(!flag) {
                fprintf(stderr, "all partitions are arrived on receiver but request is not complete, iteration %d\n", iter);
                fflush(stderr);
                errs++;
            }
```

minsii commented on test/mpi/part/start_pready.c:
```diff
@@ -0,0 +1,188 @@
+
+#define VAL(_partition, _count, _off, _iter)        \
+        (double) ((_partition) * (_count) + (_off) + (_iter))   /* Equals to `index + iter` */
+
```
replaced with `VAL(index, iter)`

minsii commented on test/mpi/part/start_pready.c:
```diff
@@ -0,0 +1,188 @@
+
+                MTestPrintfMsg(1, "Rank %d sends partition range %d:%d "
+                               "with count %ld of basic elements\n", rank, lo, high, scount);
+                MPI_Pready_list(np, arrayp, &req);
```
added the non-consecutive list version in start_pready

minsii commented on test/mpi/part/testlist:
```diff
@@ -0,0 +1,31 @@
+parrived_wait 2 arg=-spartitions=512 arg=-rpartitions=512 arg=-tot_count=4194304 arg=-iteration=5
+parrived_wait 2 arg=-spartitions=1 arg=-rpartitions=16 arg=-tot_count=64 arg=-iteration=5
+parrived_wait 2 arg=-spartitions=16 arg=-rpartitions=32 arg=-tot_count=64 arg=-iteration=5
+parrived_wait 2 arg=-spartitions=8192 arg=-rpartitions=512 arg=-tot_count=4194304 arg=-iteration=5
```
will fix

minsii commented on test/mpi/part/testlist:
```diff
@@ -0,0 +1,31 @@
+start_pready 2 arg=-spartitions=1 arg=-rpartitions=16 arg=-tot_count=64 arg=-range=1 arg=-iteration=5
+start_pready 2 arg=-spartitions=16 arg=-rpartitions=32 arg=-tot_count=64 arg=-range=2 arg=-iteration=5
+start_pready 2 arg=-spartitions=8192 arg=-rpartitions=512 arg=-tot_count=4194304 arg=-range=64 arg=-iteration=5
+
```
`count ` might be confusing as it is the total count but not per-partition-count. OK to shorten the other args.

minsii replied:
Found a bit hard to replace `iteration`->`iter`, because internally use `iter` as the current iteration id. I think it is fine to keep `iteration` as the arg name.

minsii commented:
Testing after addressing review comments for the test commits.
test:mpich/ch3/most
test:mpich/ch4/most

minsii commented:
test:mpich/ch3/most
test:mpich/ch4/most

minsii commented:
@hzhou can  you please review again? I have addressed all review comments for the tests. As you suggested, I also added two new variants: `start_pready_list_non_consecutive` and `parrived_pready_out_of_order` .

minsii commented:
The ch3/tcp failure is a timeout with RMA test. I don't think it is related to this PR.

hzhou commented:
Rerun test to confirm:

test:mpich/ch3/tcp

hzhou reviewed (commented):
The PR generally looks good. I am going to make some nitpicks next if you don't mind.

How about merge the commits from 
    "ch4: add partitioned pt2pt in netmod|shm am fallback "  to 
    "ch4/shm: add partitioned pt2pt "?
These commits are just rolling out the ch4 API and doesn't contain much complexity. I think that reducing the commits here helps vendor cherry-picking (a bit).

hzhou commented:
test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

minsii commented:
> How about merge the commits from
"ch4: add partitioned pt2pt in netmod|shm am fallback " to
"ch4/shm: add partitioned pt2pt "?
These commits are just rolling out the ch4 API and doesn't contain much complexity. I think that reducing the commits here helps vendor cherry-picking (a bit).

Addressed

minsii commented:
test:mpich/ch3/most
test:mpich/ch4/most

test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

minsii commented:
@hzhou I tried to reproduce the `threads/rma/multifence` failure reported at the [previous build](https://jenkins-pmrs.cels.anl.gov/job/mpich-review-custom/436/testReport/junit/(root)/summary_junit_xml/_____threads_rma_multifence_3__/) 

It looks like it is a bug in `MPID_Alloc_mem` when per-obj lock is enabled. I don't go deeper into this bug thus not sure what the actual cause is. But I believe it is irrelevant to this PR.

hzhou commented:
> 
> 
> @hzhou I tried to reproduce the `threads/rma/multifence` failure reported at the [previous build](https://jenkins-pmrs.cels.anl.gov/job/mpich-review-custom/436/testReport/junit/(root)/summary_junit_xml/_____threads_rma_multifence_3__/)
> 
> It looks like it is a bug in `MPID_Alloc_mem` when per-obj lock is enabled. I don't go deeper into this bug thus not sure what the actual cause is. But I believe it is irrelevant to this PR.
> 
> Have you seen this bug before?

I don't think I have. I can take a look if it re-occurs.

minsii commented:
It happens about once per 10 runs on my local build.

hzhou commented:
> It happens about once per 10 runs on my local build.

Thanks. It might be introduced by https://github.com/pmodels/mpich/pull/4746. I'll take care of it.

minsii commented:
@hzhou Great. Will you make another round of review? I have addressed all posted comments.

hzhou commented:
> 
> 
> @hzhou I tried to reproduce the `threads/rma/multifence` failure reported at the [previous build](https://jenkins-pmrs.cels.anl.gov/job/mpich-review-custom/436/testReport/junit/(root)/summary_junit_xml/_____threads_rma_multifence_3__/)
> 
> It looks like it is a bug in `MPID_Alloc_mem` when per-obj lock is enabled. I don't go deeper into this bug thus not sure what the actual cause is. But I believe it is irrelevant to this PR.
> 
> Have you seen this bug before?

@minsii I won't have time to investigate the bug in a few days. Could you file a issue for it and attach test logs?

Meanwhile, I'm going to approve this PR.

hzhou reviewed (approved):
@minsii PR looks good. Thanks for the work. One last request, could you in the PR description add some comments on the architecture of this PR? What I am looking for is a list of new files this PR added and short descriptions of their purpose, as well as some overview of the rest of the changes. The PR is spread over 26 commits 3000 loc, so I think having an overview can assist future development.

hzhou commented:
> I won't have time to investigate the bug in a few days. Could you file an issue for it and attach test logs?

Never mind, I found the bug, will push a PR to fix it soon (independent of this PR).

minsii commented:
Last time trigger tests after rebasing over latest main branch

test:mpich/ch3/most
test:mpich/ch4/most

test:mpich/custom --enable-thread-cs=per-vci --with-ch4-max-vcis=8
netmod: ch4:ucx
env: MPIR_CVAR_CH4_NUM_VCIS=4

minsii commented:
@hzhou I updated the top comment. Can you please see if the **[List of major changes/new files]** section addresses your request?
https://github.com/pmodels/mpich/pull/5033#issue-562803755

hzhou commented:
> @hzhou I updated the top comment. Can you please see if the **[List of major changes/new files]** section addresses your request?
> [#5033 (comment)](https://github.com/pmodels/mpich/pull/5033#issue-562803755)

Looks good. Thanks
